skxray/correlation.py

```diff
@@ -0,0 +1,375 @@
+    g2 = np.zeros((num_levels, num_qs), dtype=np.float64)
+
+    # correlation for delays, images must be keep for up to maximum
+    # delay in buf
+    buf = np.zeros((num_levels, num_bufs, np.sum(num_pixels)),
+                   dtype=np.float64)
+
+    # to track processing each level
+    cts = np.zeros(num_levels)
+
+    # to increment buffer
+    cur = np.ones(num_levels, dtype=np.int64)
+
+    # to track how many images processed in each level
+    num = np.zeros(num_levels, dtype=np.int64)
+
+    # starting time for the process
+    t1 = time.time()
+
+    for n, img in enumerate(img_stack):  # changed the number of frames
+
+        cur[0] = (1 + cur[0]) % num_bufs  # increment buffer
+
+        # add image data to the buf to use for correlation
+        buf[0, cur[0] - 1] = (np.ravel(img))[pixel_list]
+
+        # call the _process function for multi-tau level one
+        G, IAP, IAF, num = _process(buf, G, IAP, IAF, q_inds,
+                                    num_bufs, num_pixels, num, level=0,
+                                    buf_no=cur[0] - 1)
+
+        # check whether the number of levels is one, otherwise
+        # continue processing the next level
+        if num_levels > 1:
+            processing = True
+        else:
+            processing = False
+
+        # the image data will be saved in buf according to each level then call
+        #  _process function to calculate one time correlation functions
+        level = 1
+        while processing is True:
+            if cts[level]:
+                prev = 1 + (cur[level - 1] - 2 )%num_bufs
+                cur[level] = 1 + cur[level]%num_bufs
+
+                buf[level, cur[level] - 1] = (buf[level - 1, prev - 1] +
+                                              buf[level - 1,
+                                                  cur[level - 1] - 1])/2
+
+                # make the cts zero once that level is processed
+                cts[level] = 0
+
+                # call the _process function for each multi-tau level
+                # for multi-tau levels greater than one
+                G, IAP, IAF, num = _process(buf, G, IAP, IAF, q_inds,
+                                       num_bufs, num_pixels, num,
+                                       level=level, buf_no=cur[level]-1,)
+                level += 1
+
+                # Checking whether there is next level for processing
+                if level<num_levels:
+                    processing = True
+                else:
+                    processing = False
+            else:
+                cts[level] = 1
+                processing = False
+
+    # ending time for the process
+    t2 = time.time()
+
+    logger.info("Processing time for {0} images took {1} seconds."
+                "".format(len(img_stack), (t2-t1)))
+
+    # to get the final G, IAP and IAF values
+    g_max = IAP.shape[0]
+
+    # calculate the one time correlation
+    g2 = (G[: g_max] / (IAP[: g_max] * IAF[: g_max]))
```

Comment:
I don't understand the slicing logic here, they should all be the same shape right?